File: Insertion_Sort/Insertion_Sort.rb
#Function for insertion sort
def insertionSort (array)
  for i in 1 .. array.length - 1
    temp = array[i]                 #Store key value
    j = i - 1
    while j > 0 and array[j] > temp #Comparing with key and swapping accordingly
      array[j+1] = array[j]
      j = j - 1
    end
    array[j+1] = temp               #Final swap
  end
end

#Hardcoded array
array = [0,9,1,4,2]

#Pre sort array
print(array)

puts("\nAfter Sorting:\n")

insertionSort(array)

#Sorted Array
print(array)

#Output
#[0, 9, 1, 4, 2]
#[0, 1, 2, 4, 9]

File: Insertion_Sort/Insertion_Sort.java
import java.util.Scanner;

class Insertion_Sort
{
    // function for insertion sort
    public static void InsertionSort(int[] array, int size)
    {
        int temp, j;

        for(int i = 1; i < size; i++)
        {
            temp = array[i];
            j = i - 1;

            // Do swapping
            while(j >= 0 && array[j] > temp)
            {
                array[j + 1] = array[j];
                j--;
            }

            array[j + 1] = temp;
        }
    }

    // function ro print array
    public static void Print_Array(int[] array, int size)
    {
        for(int i = 0; i < size; i++)
            System.out.print(array[i] + " ");

        System.out.println();
    }

    public static void main(String[] args)
    {
        int num;
        Scanner s = new Scanner(System.in);
        num = s.nextInt();
        int array[] = new int[num];
        for (int i = 0; i < num; i++) {
            array[i] = s.nextInt();
        }
        InsertionSort(array, num);
        Print_Array(array, num);
    }
}

/*
Input :
num = 6
array = [1, 6, 3, 3, 5, 2]

Output :
[1, 2, 3, 3, 5, 6]
*/
